aptos-move/block-executor/src/executor.rs

```diff
@@ -539,25 +535,79 @@ where
                                 finalized_group.is_empty(),
                                 metadata_op.is_deletion()
                             );
-                            maybe_err = Some(Error::FallbackToSequential(PanicOr::Or(
+                            return Err(Error::FallbackToSequential(PanicOr::Or(
                                 IntentionalFallbackToSequential::ResourceGroupError,
                             )));
                         }
-                        finalized_groups.push((group_key, metadata_op, finalized_group));
+                        Ok((group_key, metadata_op, finalized_group))
                     },
                     Err(e) => {
                         error!("Error committing resource group {:?}", e);
-                        maybe_err = Some(Error::FallbackToSequential(PanicOr::Or(
+                        Err(Error::FallbackToSequential(PanicOr::Or(
                             IntentionalFallbackToSequential::ResourceGroupError,
-                        )));
+                        )))
                     },
-                };
+                }
+            };
 
+            let group_metadata_ops = last_input_output.group_metadata_ops(txn_idx);
+            let mut finalized_group_writes = Vec::with_capacity(group_metadata_ops.len());
+            let mut maybe_err = None;
+            for (group_key, metadata_op) in group_metadata_ops.into_iter() {
+                // finalize_group copies Arc of values and the Tags (TODO: optimize as needed).
+                match finalize_group(group_key, metadata_op) {
+                    Ok((group_key, metadata_op, finalized_group)) => {
+                        finalized_group_writes.push((group_key, metadata_op, finalized_group));
+                    },
+                    Err(err) => {
+                        maybe_err = Some(err);
+                        break;
+                    },
+                }
                 if maybe_err.is_some() {
                     break;
                 }
             }
-            last_input_output.record_finalized_group(txn_idx, finalized_groups);
+
+            if maybe_err.is_none() {
+                if let Some(group_reads_needing_delayed_field_exchange) =
+                    last_input_output.group_reads_needing_delayed_field_exchange(txn_idx)
+                {
+                    for (group_key, metadata_op) in
+                        group_reads_needing_delayed_field_exchange.into_iter()
+                    {
+                        match finalize_group(group_key, metadata_op) {
+                            Ok((group_key, metadata_op, finalized_group)) => {
+                                if let Some(metadata_op) =
+                                    metadata_op.convert_read_to_modification()
```

Comment:
Let's do the conversion not here, but when we construct the write-op (metadata_op), because it is used for gas charging which matches on the op kind. I think currently storage refund may be affected, but even if it's not, later some specialized logic may be introduced for creation vs modification and things would break.
This also applies to read resources containing delayed fields that we convert to writes, I think, @georgemitenkov - so may be worth doing as a follow-up, but we should then add a agg_v2 TODO for sure.

Reply:
Done.